Userland/Libraries/LibWeb/Animations/Animation.cpp

```diff
@@ -368,6 +437,160 @@ WebIDL::ExceptionOr<void> Animation::silently_set_current_time(Optional<double>
+            // 5. Set finishEvent’s currentTime attribute to the current time of animation.
+            auto& realm = this->realm();
+            AnimationPlaybackEventInit init;
+            init.current_time = current_time();
+            auto finish_event = heap().allocate<AnimationPlaybackEvent>(realm, realm, "finish"_fly_string, init);
+
+            // 6. Set finishEvent’s timelineTime attribute to the current time of the timeline with which animation is
+            //    associated. If animation is not associated with a timeline, or the timeline is inactive, let
+            //    timelineTime be null.
+            if (m_timeline && !m_timeline->is_inactive())
+                finish_event->set_timeline_time(m_timeline->current_time());
+            else
+                finish_event->set_timeline_time({});
+
+            // 7. If animation has a document for timing, then append finishEvent to its document for timing's pending
+            //    animation event queue along with its target, animation. For the scheduled event time, use the result
+            //    of converting animation’s associated effect end to an origin-relative time.
+            if (auto document_for_timing = this->document_for_timing()) {
+                document_for_timing->append_pending_animation_event({ .event = finish_event,
+                    .target = *this,
+                    .scheduled_event_time = convert_a_timeline_time_to_an_origin_relative_time(associated_effect_end()) });
+            }
+            //    Otherwise, queue a task to dispatch finishEvent at animation. The task source for this task is the DOM
+            //    manipulation task source.
+            else {
+                HTML::queue_global_task(HTML::Task::Source::DOMManipulation, realm.global_object(), [this, finish_event]() {
+                    dispatch_event(finish_event);
+                });
+            }
+
+            m_has_finish_notification_microtask_scheduled = false;
+        };
+
+        // 2. If synchronously notify is true, cancel any queued microtask to run the finish notification steps for this
+        //    animation, and run the finish notification steps immediately.
+        if (synchronously_notify == SynchronouslyNotify::Yes) {
+            m_should_abort_finish_notification_microtask = false;
+            finish_notification_steps();
+            m_should_abort_finish_notification_microtask = true;
+        }
+        //    Otherwise, if synchronously notify is false, queue a microtask to run finish notification steps for
+        //    animation unless there is already a microtask queued to run those steps for animation.
+        else {
+            if (!m_has_finish_notification_microtask_scheduled)
+                HTML::queue_a_microtask({}, move(finish_notification_steps));
+
+            m_has_finish_notification_microtask_scheduled = true;
+            m_should_abort_finish_notification_microtask = false;
+        }
```

Comment:
I'd love if someone could double-check my logic here with these boolean flags. The idea is that the finish notification microtask should never be scheduled more than once. If we run this with `SynchronouslyNotify::Yes`, then we should run the steps immediately and cancel the pending microtask. If we run with `SynchronouslyNotify::No`, then we should schedule the microtask, but only if it is not already scheduled. 

Afaik microtasks can't be removed from the event loop once they're scheduled, so I use these boolean flags to track the state. It should _hopefully_ be self explanatory; if not, I'll need to add some clarifying comments.